### PR TITLE
[JUJU-3213] DeployFromRepository validate endpoint bindings, placement and number of units

### PR DESCRIPTION
**Pull request by hmlanigan**
A follow on to #15278 

Now validating endpoint bindings, placement and num units. The first two should behave the same as a traditional charm deploy now. The latter will behave the same with the juju cli, but a little different with direct usable of the facade call. 

Today, number of units defaults to 1 if not specified when using the juju cli. As DeployFromRepository given a single argument value of the charm name, should deploy a single unit of the charm, ensure that if not provided, the number of units deployed is 1.

Regarding placement, there is currently a bug in 2.9+ where juju deploy to an existing machine may or may not find the correct operating system to use for the charm depending on luck or the series being specific on the cli. There is an 
attempt here to fix that problem. However there is still a timing window between add-machine and deploy where results maybe non deterministic. The new code relies on hardware characteristics of a deployed machine to determine the operating system for the charm. However if the hardware characteristics have not yet been set, juju may not deploy with the correct charm. There may be a way to resolve this, created a new Jira ticket: https://warthogs.atlassian.net/browse/JUJU-3279

## QA steps

```sh
$ JUJU_DEV_FEATURE_FLAGS=server-side-charm-deploy juju bootstrap localhost test
$ juju add-model five

# juju deploy ubuntu defaulted to 20.04 for me, perhaps a bug, so add a machine of a different operating system
# to see that juju deploy chooses the operating system for charm deployment from the machine specified.
$ juju add-machine ubuntu@22.04
created machine 0

# There is a timing window to be resolved here, where juju may not pick up the correct operating system if 
# deploy is called too quickly after add-machine. This behavior is an improvement over current behavior but 
# could use fixing, though it'd be complicated.
$ JUJU_DEV_FEATURE_FLAGS=server-side-charm-deploy juju deploy ubuntu --to jammy-ubuntu
application.DeployInfo{
    CharmURL:     "ch:amd64/jammy/ubuntu-21",
    Channel:      "stable",
    Architecture: "amd64",
    Base:         series.Base{
        OS:      "ubuntu",
        Channel: series.Channel{Track:"22.04", Risk:"stable"},
    },
    EffectiveChannel: (*string)(nil),
}

# deploy with a new container and see machine 1/lxd/0 created.
$ JUJU_DEV_FEATURE_FLAGS=server-side-charm-deploy juju deploy ubuntu --to lxd container-ubuntu
application.DeployInfo{
    CharmURL:     "ch:amd64/focal/ubuntu-21",
    Channel:      "stable",
    Architecture: "amd64",
    Base:         series.Base{
        OS:      "ubuntu",
        Channel: series.Channel{Track:"20.04", Risk:"stable"},
    },
    EffectiveChannel: (*string)(nil),
}

# create machines with different operating systems and deploy, should fail as the operating systems do not match.
$ juju add-machine --base ubuntu@22.04 ; juju add-machine --base ubuntu@20.04
created machine 3
created machine 4
$ JUJU_DEV_FEATURE_FLAGS=server-side-charm-deploy juju deploy ubuntu --to 3,4 -n 2
application.DeployInfo{}
ERROR bases of existing placement machines do not match

# $ JUJU_DEV_FEATURE_FLAGS=server-side-charm-deploy juju deploy ubuntu five -n 3
application.DeployInfo{
    CharmURL:     "ch:amd64/focal/ubuntu-21",
    Channel:      "stable",
    Architecture: "amd64",
    Base:         series.Base{
        OS:      "ubuntu",
        Channel: series.Channel{Track:"20.04", Risk:"stable"},
    },
    EffectiveChannel: (*string)(nil),
}
$ juju status
Model  Controller  Cloud/Region         Version      SLA          Timestamp
test   test        localhost/localhost  3.2-beta1.1  unsupported  21:12:00Z

App     Version  Status   Scale  Charm   Channel  Rev  Exposed  Message
five             waiting    0/3  ubuntu  stable    21  no       waiting for machine

# test bindings
$ JUJU_DEV_FEATURE_FLAGS=server-side-charm-deploy juju bootstrap aws test-aws
$ juju add-model testing
# double check the subnets, then create a space.
$ juju add-space testing
$ JUJU_DEV_FEATURE_FLAGS=server-side-charm-deploy juju deploy cinder --bind "shared-db=testme"
application.DeployInfo{
    CharmURL:     "ch:amd64/xenial/cinder-530",
    Channel:      "stable",
    Architecture: "amd64",
    Base:         series.Base{
        OS:      "ubuntu",
        Channel: series.Channel{Track:"20.04", Risk:"stable"},
    },
    EffectiveChannel: (*string)(nil),
}
$ juju show-application cinder | grep shared-db
    shared-db: testme
```

### CODE REVIEW

**Review by cderici — APPROVED** [review]
QA went well. Code LGTM.

File: apiserver/facades/client/application/application.go [hunk 1]
```diff
@@ -731,16 +731,19 @@ func parseCharmSettings(modelType state.ModelType, ch Charm, appName string, cfg
 	return appConfig, appCfgSchema, charmSettings, schemaDefaults, nil
 }
 
+type MachinePlacementBackend interface {
```
[the remark this answers is not in the document]
Comment (hmlanigan) [reply apiserver/facades/client/application/application.go:734]: The DeployFromRepositoryAPI will also use this method and does not have nor need the entire application.Backend. As well, the function requiring it, only needs the one Machine method as well, not the entire backend.

File: apiserver/facades/client/application/application.go [hunk 2]
```diff
@@ -731,16 +731,19 @@ func parseCharmSettings(modelType state.ModelType, ch Charm, appName string, cfg
...
+	Machine(string) (Machine, error)
+}
+
 // checkMachinePlacement does a non-exhaustive validation of any supplied
 // placement directives.
 // If the placement scope is for a machine, ensure that the machine exists.
 // If the placement is for a machine or a container on an existing machine,
 // check that the machine is not locked for series upgrade.
-func checkMachinePlacement(backend Backend, args params.ApplicationDeploy) error {
+func checkMachinePlacement(backend MachinePlacementBackend, app string, placement []*instance.Placement) error {
```
[the remark this answers is not in the document]
Comment (hmlanigan) [reply apiserver/facades/client/application/application.go:743]: I see your point. However I'm going to keep this consistent will all of the other variable names for a `[]*instance.Placement` in juju as placement without the s.

File: apiserver/facades/client/application/application.go [hunk 3]
```diff
@@ -518,7 +518,7 @@ func deployApplication(
 
 	// This check is done early so that errors deeper in the call-stack do not
 	// leave an application deployment in an unrecoverable error state.
-	if err := checkMachinePlacement(backend, args); err != nil {
+	if err := checkMachinePlacement(backend, args.ApplicationName, args.Placement); err != nil {
```
Comment (manadart) [inline apiserver/facades/client/application/application.go:521]: I'd tend not to expand this, fewer argument being better; also less likely to require change in future.

File: apiserver/facades/client/application/deployrepository.go [hunk 4]
```diff
@@ -273,6 +291,22 @@ func (v *deployFromRepositoryValidator) validate(arg params.DeployFromRepository
 		placement:       arg.Placement,
 		storage:         stateStorageConstraints(arg.Storage),
 	}
+	if arg.NumUnits != nil {
```
Comment (manadart) [inline apiserver/facades/client/application/deployrepository.go:294]: I'd be inclined to break this method up if it needs further additions.

**Pull request merged**